### ISSUE TITLE
"__str__ does not return str (Pylint E0307)" while it does return a str

### ISSUE THREAD

**Issue by StickyMouse27** [PyCQA/pylint#10640]
### Bug description

```python
# pylint: disable=missing-docstring


class A:
    def __new__(cls, name: str) -> str:
        return name


class B:
    def __str__(self) -> str:
        return A("B")


print(B())
```

### Configuration

```ini

```

### Command used

```shell
pylint test.py
```

### Pylint output

```python
************* Module test
test.py:4:0: R0903: Too few public methods (1/2) (too-few-public-methods)
test.py:10:4: E0307: __str__ does not return str (invalid-str-returned)
test.py:9:0: R0903: Too few public methods (1/2) (too-few-public-methods)

-----------------------------------
Your code has been rated at 0.00/10
```

### Expected behavior

A("B") does return a str, so there shouldn't be an error

### Pylint version

```shell
pylint 3.3.9
astroid 3.3.11
Python 3.12.10 (tags/v3.12.10:0cc8128, Apr  8 2025, 12:21:36) [MSC v.1943 64 bit (AMD64)]
```

### Additional dependencies

```python

```

**Comment by jacobtylerwalls** [comment]
Duplicate of #7258
